### ISSUE TITLE
Recipe search api

### ISSUE THREAD

**Issue by everycook** [everycook/EveryCook#185]
We should be able to search for recipes using the following keywords:
* lang= for language
* title= to search the titles for a string
* w_ing[]= with ingredient (can be used multiple times)
* wo_ing[]= with ingredient (can be used multiple times)

and get a json output containing:
* lang
* title
* img_url
* rec_id
* and probably some other values I don't think of right now.

We want to include in every api call:
* token (the token to allow access, as long as we have no ssl, "everycook" is the only accepted token

We allow only read access by now, to void conflicts. later we want write access to some parts and have to ckeck if the token is safe enough for this if we have ssl active

**Comment by tbuerki** [comment]
Search can also go via rec_type (the type of the recipes) and rec_cuisine (the recipes cuisine). The output is a list as suggested in Issue #186 that can be merged into this one (same API call)

**Comment by swerder** [comment]
first version done.
url is api/searchRecipes

possible parameters are currently:
- 'lang', String: language to use, if no valid one EN_GB is used.
- 'searchEngine', String('yii'(=default), 'solr'): choose search engine to use, currently only yii is implemented
- 'title', String: search in the title of the queried language.
- 'query', String: will be searched over all fields (only on "solr" searchEngine, on "yii" engine same as title
- 'start_with', String: title or synonyme start with
- 'difficulty', int, difficulty of recipe (queried by dif_id)
- 'rating', float[but currently only int], rating of recipe (queried by exact value)
- 'rating_above', float[but currently only int], rating of recipe (queried by value higher or equal submitted value)
- 'rating_below', float[but currently only int], rating of recipe (queried by value lower submitted value)
- 'w_ing[]', int, can be used multiple times: recipe with this ingredient (queried by ing_id)
- 'w_ing', int list separated with ',': recipe with this ingredient (queried by ing_id)
- 'wo_ing[]', int, can be used multiple times: recipe without this ingredient (queried by ing_id)
- 'wo_ing', int list separated with ',': recipe without this ingredient (queried by ing_id)
- 'rec_type', int: recipe from this type (queried by ret_id)
- 'rec_cuisine', int: recipe with this cusine (queried by cut_id)
- 'rec_sub_cuisine', int: recipe with this cusine (queried by cst_id)
- 'rec_sub_sub_cuisine', int: recipe with this cusine (queried by css_id)
- 'tag', string: recipe with this tag (queried by language dependent tag text)
- 'rec_id[]', int, can be used multiple times: recipe with specified id
- 'rec_id', int list separated with ',': recipe with specified id

All parameters are optional, if not limited, all recipes are returned in lang EN_GB

additional there are paging parameters:
- 'offset', int: start offset/index for paging, default 0
- 'limit', int: amount of results for page, default 10
- 'sort', String(value list): sort order
  - 'score'=> score
  - 'az'=>name ascending
  - 'za'=>name descending
  - 'new'=>last change DESC
  - 'old'=>last change 
  - 'K'=>kcal / portion
  - 'k'=>kcal / portion DESC
  - 'D'=>difficulty
  - 'd'=>difficulty DESC
  - 'P'=>prepare time
  - 'p'=>prepare time DESC
  - 'C'=>cook time,
  - 'c'=>cook time DESC
  - 'T'=>total time
  - 't'=>total time DESC
  - 'R'=>rating
  - 'r'=>rating DESC

**Comment by everycook** [comment]
what's the url? can you make an example call link? thx

**Comment by swerder** [comment]
full url example:
http://everycook.org/db_wsd/api/searchRecipes?token=everycook&lang=de_ch&w_ing=1&query=ris

**Comment by everycook** [comment]
If I want to make a list of recipe with the first 10, then 11-20 and so on. is that implemented?

**Comment by everycook** [comment]
In the apache error log I get something like:
it means I can't get images from dizmo, but it works in browser...

[Wed Mar 11 14:07:28.623286 2015] [:error] [pid 15111] [client 31.164.183.236:46999] PHP Fatal error:  Using $this when not in object context in /var/www/db_wsd/protected/components/Controller.php on line 131, referer: http://127.0.0.1/dizmo
[Wed Mar 11 14:07:28.726709 2015] [:error] [pid 15096] [client 31.164.183.236:47000] PHP Fatal error:  Using $this when not in object context in /var/www/db_wsd/protected/components/Controller.php on line 131, referer: http://127.0.0.1/dizmo

**Comment by swerder** [comment]
image bug fixes.
paging & sort parameter added in description above.

**Comment by swerder** [comment]
from everycook in 186:
it would be great to have a parameter "lastchange" so we can check if a cached recipe is still up to date or if we have to re-sync it.

**Comment by swerder** [comment]
added lastchange field as unixtimestamp.
attention: it is the last change of recipe and/or steps, but if someone change the options of an ingredient this would not reflect in this value.

**Comment by everycook** [comment]
We also need subcuisine and subsubcuisine types.

**Comment by everycook** [comment]
If I make:
http://everycook.org/db_wsd/api/searchRecipes?token=everycook&lang=de_ch&rec_cuisine=4
I get the titles. but in:
http://everycook.org/db_wsd/api/searchRecipes?token=everycook&lang=de_ch&rec_cuisine=2
The titles are empty

**Comment by everycook** [comment]
The bug is probably due to incomplete recipes in db_wsd. because it does not appear in db.

**Comment by swerder** [comment]
correct in language en_gb there would be the titles:
http://everycook.org/db_wsd/api/searchRecipes?token=everycook&lang=en_gb&rec_cuisine=2
title is only in en_gb a required field, other languages can be empty.

**Comment by everycook** [comment]
I need a field tags:{1,2,3} in the list of results

**Comment by everycook** [comment]
We should not get results twice. If I look for two tags and a recipe has these two tags I get it twice in the results.

**Comment by swerder** [comment]
tags added (as array [1,2,3] not object {1,2,3})
always return distinct results (each entry only once)

**Comment by everycook** [comment]
we also need:
Preparation time
Cooking time
difficulty
rating (if available)

**Comment by everycook** [comment]
nice, now please give me a parameter id (or ids) where I can give several rec ids to get a list of those. I need this to retrieve a list of locally stored favorites.

**Comment by swerder** [comment]
There is now a new parameter rec_id, to search for recipes by id.

**Comment by everycook** [comment]
doesn't work for me. if I call http://everycook.org/db/api/searchRecipes?token=everycook&rec_id=19,15,17&offset=0&limit=1000&ver=1&lang=de_ch I get 29 results, I want 3

**Comment by swerder** [comment]
but now it works.

**Comment by everycook** [comment]
sorting by difficulty and rating would be great

**Comment by swerder** [comment]
Added parameter:
- 'difficulty', int, difficulty of recipe (queried by dif_id)
- 'rating', float[but currently only int], rating of recipe (queried by exact value)
- 'rating_above', float[but currently only int], rating of recipe (queried by value higher or equal submitted value)
- 'rating_below', float[but currently only int], rating of recipe (queried by value lower submitted value)

Updated sort param: add difficulty, rating and cook time.

Updated the param list in comment above.

**Comment by everycook** [comment]
i get a Unknown column 'CANGED_ON' in 'order clause'
http://everycook.org/dizmo/api/searchRecipes?token=everycook&offset=0&limit=1000&ver=1&lang=de_ch&sort=new&rec_type=3&query=riso

**Comment by swerder** [comment]
2 errors: typo in field name and it's aliased and therefore not found anyway...
fixed
